// @octokit/webhooks/dist-src/index.js
import { createEventHandler } from "./event-handler/index";
import { createMiddleware } from "./middleware/index";
import { middleware } from "./middleware/middleware";
import { verifyAndReceive, } from "./middleware/verify-and-receive";
import { sign } from "./sign/index";
import { verify } from "./verify/index";
// U holds the return value of `transform` function in Options
class Webhooks {
    constructor(options) {
        if (!options || !options.secret) {
            throw new Error("[@octokit/webhooks] options.secret required");
        }
        const state = {
            eventHandler: createEventHandler(options),
            path: options.path || "/",
            secret: options.secret,
            hooks: {},
        };
        this.sign = sign.bind(null, options.secret);
        this.verify = verify.bind(null, options.secret);
        this.on = state.eventHandler.on;
        this.onAny = state.eventHandler.onAny;
        this.onError = state.eventHandler.onError;
        this.removeListener = state.eventHandler.removeListener;
        this.receive = state.eventHandler.receive;
        this.middleware = middleware.bind(null, state);
        this.verifyAndReceive = verifyAndReceive.bind(null, state);
    }
}
const createWebhooksApi = Webhooks.prototype.constructor;
export { createEventHandler, createMiddleware, createWebhooksApi, Webhooks, sign, verify, };
